File: PLeetcode/com/cx363/algorithm/QuickSort.py
'''
QuickSort Demo

Created on Oct 20, 2014

@author: xiaocunqi
'''
import random
class QuickSort:
    DESCENT = 1
    ASCENT  = 0
    def sort(self, data, order = 0):
        rand = random.Random()
        low = 0
        high = len(data)
        
        self.__quickSort(data, order, low, high - 1, rand)
        
        return data
    
    def __quickSort(self, data, order, low, high, rand):
        if high - 1 <= low:
            pass
        else:
            sentry_index = rand.randrange(low, high)
            
            sentry = data[sentry_index]
            
            self.__swipe(data, sentry_index, high)
            
            index = low
            for i in range(low, high):
                
                if order == QuickSort.ASCENT:
                    if data[i] <= sentry:
                        self.__swipe(data, i, index)
                        index += 1
                else:
                    if data[i] >= sentry:
                        self.__swipe(data, i, index)
                        index += 1
                
                
            self.__swipe(data, index, high)
            sentry_index = index
                
            self.__quickSort(data, order, low, sentry_index - 1, rand)
            self.__quickSort(data, order, sentry_index + 1, high, rand)
    
    def __swipe(self, data, source, target):
        tmp = data[source]
        data[source] = data[target]
        data[target] = tmp
        
            
quickSort = QuickSort()
numbers = [1, 5, 3, 77, 9, 2, 10, 7, 8]

print(quickSort.sort(numbers))
            
                    
            

File: Leetcode/src/com/cx363/algorithm/QuickSort.java
package com.cx363.algorithm;

import java.util.Random;

public class QuickSort {
	public static final int ASCENT = 0;
	public static final int DESCENT = 1;
	
	public int[] sort(int[] data, int order){
		int low = 0;
		int high = data.length;
		
		Random random = new Random(System.currentTimeMillis());
		
		quickSort(data, low, high - 1, order, random);
		return data;
	}
	
	private void quickSort(int[] data, int low, int high, int order, Random random){
		if(low < high){
			int sentry = partition(data, low, high, order, random);
			quickSort(data, low, sentry-1, order, random);
			quickSort(data, sentry+1, high, order, random);
		}
	}
	
	private int partition(int[] data, int low, int high , int order, Random random){
		int p = low + random.nextInt(high - low + 1);	//important
		int sentry = data[p];
		
		swipe(data, p, high);
		
		int index = low;
		
		for(int i=low; i<high; i++){
			if(QuickSort.ASCENT == order){
				if(data[i] <= sentry){
					swipe(data, index, i);
					index++;
				}
			}else{
				if(data[i] >= sentry){
					swipe(data, index, i);
					index++;
				}
			}
		}
		
		swipe(data, index, high);
		
		
		return index;
	}
	
	private void swipe(int[] data, int source, int target){
		int temp = data[source];
		data[source] = data[target];
		data[target] = temp;
	}
	
	public static void main(String[] args){
		int[] data = {2, 1, 3, 5, 8, 19, 77, 4, 9};
		
		QuickSort quickSort = new QuickSort();
		
		int[] result = quickSort.sort(data, QuickSort.ASCENT);
		
		for(int a : result){
			System.out.print(a+",");
		}
	}
}
